{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "92092038",
   "metadata": {},
   "source": [
    "# Dimensionality reduction, clustering and annotation"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "95cde18a",
   "metadata": {},
   "source": [
    "In this notebook, we will cover the topics of: \n",
    "- Dimensionality reduction \n",
    "- Clustering\n",
    "- Annotation"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1c772fb6",
   "metadata": {},
   "source": [
    "# Library imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69e629d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n",
    "import os\n",
    "\n",
    "# single cell library\n",
    "import scanpy as sc\n",
    "\n",
    "# data loading\n",
    "import lamindb as ln"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2970f5d6",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "cc3321af",
   "metadata": {},
   "source": [
    "We read the data that has already been pre-processed, i.e., QC and normalized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "091ece7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ln.Artifact.df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b5e9b52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the artifact\n",
    "artifact = ln.Artifact.get(\"7YRPApRIFl4LOsaj0000\")\n",
    "\n",
    "# Load the AnnData object into memory\n",
    "adata = artifact.load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1045dca3",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_cols_to_keep = [\n",
    "    'sample', 'condition', 'area', 'age', 'donor', \n",
    "    'sequencing batch', 'sequencing id', 'sex', \n",
    "    'APOE genotype', 'brain area number', 'running number', \n",
    "    'donor_v2', 'donor_v2_area', 'tech_sample'\n",
    "]\n",
    "\n",
    "var_cols_to_keep = ['ensemble_id', 'Accession', 'Chromosome', 'End', 'Start', 'Strand']\n",
    "\n",
    "# Delete unwanted obs columns\n",
    "obs_cols_to_delete = [col for col in adata.obs.columns if col not in obs_cols_to_keep]\n",
    "for col in obs_cols_to_delete:\n",
    "    del adata.obs[col]\n",
    "\n",
    "# Delete unwanted var columns\n",
    "var_cols_to_delete = [col for col in adata.var.columns if col not in var_cols_to_keep]\n",
    "for var in list(var_cols_to_delete):\n",
    "    del adata.var[var]\n",
    "\n",
    "# Delete unwanted layers\n",
    "layers_cols_to_delete = [col for col in adata.layers if col not in ['raw']]\n",
    "for layers in list(layers_cols_to_delete):\n",
    "    del adata.layers[layers]\n",
    "\n",
    "# Delete all other slots\n",
    "del adata.uns\n",
    "del adata.obsm\n",
    "del adata.varm\n",
    "del adata.obsp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f17edba-a677-4e3f-baa2-cd564595d686",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fca5484",
   "metadata": {},
   "source": [
    "# Highly Variable Genes"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "326e37ac",
   "metadata": {},
   "source": [
    "We extract highly variable genes (HVGs) to further reduce the dimensionality of the dataset and include only the most informative genes. Genes that vary substantially across the dataset are informative of the underlying biological variation in the data. As we only want to capture biological variation in these genes, we select highly variable genes after normalization and batch correction. HVGs are used for clustering, trajectory inference, and dimensionality reduction/visualization, while the full data set is used for computing marker genes, differential testing, cell cycle scoring, and visualizing expression values on the data.\n",
    "\n",
    "Typically between 1000 and 5000 genes are selected. Here, we extract the top 4000 most variable genes for further processing. If particular genes of importance are known, one could assess how many highly variable genes are necessary to include all, or the majority, of these."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fca2946d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(\n",
    "    adata, layer=\"raw\", n_top_genes=4000, flavor=\"seurat\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ac44c10",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6e7f1e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var[\n",
    "    [\"highly_variable\", \"highly_variable_rank\", \"means\", \"variances\", \"variances_norm\"]\n",
    "].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5c2bcb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.highly_variable_genes(adata)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "af444fa2",
   "metadata": {},
   "source": [
    "The plots show how the data was normalized to select highly variable genes irrespective of the mean expression of the genes. This is achieved by using the index of dispersion which divides by mean expression, and subsequently binning the data by mean expression and selecting the most variable genes within each bin."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2382818",
   "metadata": {},
   "source": [
    "# Dimensionality reduction: PCA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e5c28e0",
   "metadata": {},
   "source": [
    "scRNA-seq attempts to measure the mRNA transcripts of all the genes (~20,000) at single-cell level resolution resulting in a high-dimensional data. The cells represent the samples while the genes represent the features or dimensions of the data. These features/dimensions define the phenotype of each single cell. A cellular phenotype vector points to the location of a cell in a high-dimensional space whose each axis is represented by each of the measured features (genes).\n",
    "\n",
    "Meanwhile, we know that the genes and their products inside cells form modules, a result of genes being co-regulated. The modular and co-regulatory structure of gene interaction means that a typical scRNA-seq data displays correlation structure among groups of genes. Thus, we are capturing redundant information by measuring all genes from a single cell. Put differently, there are only a few key variables (possibly latent) that are driving the underlying biological process. One consequence of this realization is that the cellular phenotype space is really a low-dimensional object embedded onto a high-dimensional space, i.e. the biological system really has few free variables but is projected onto a space of thousands of features. We call the low-dimensional object the data manifold. It refers to the true structure of the data irrespective of how it is embedded; and understanding the manifold holds the key to uncovering the underlying biology. The notion of dimensionality reduction is thus to extract those few meaningful dimensions that are reflective of the true underlying biological process.\n",
    "\n",
    "Futhermore, performing dimensionality reduction on the data reduces the number of dimensions, so it has an added advantage of speeding-up computation (especially when dealing with millions of cells)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e8b9ec3",
   "metadata": {},
   "source": [
    "PCA (Principal Component Analysis) is one of the most popular dimensionality reduction algorithms. In short, PCA takes in a high-dimensional data as input and outputs a transformed data with the same number of samples (cells) but with reduced number of dimensions such that the recovered dimensions represent directions of highest variances in the original data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7809e67c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.pca(adata, n_comps=100, use_highly_variable=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "596be014",
   "metadata": {},
   "source": [
    "A brief explanation of the code above:\n",
    "\n",
    "We provide adata as an input to the method. As with any other Scanpy method, since we do not specify anything else, the PCA is run on adata.X. Now, we recall that adata.X holds the log-normalized data (or the z-scored normalized data if you succesfully executed sc.pp.scale function above and did not set adata.X to be log-norm data). Remember: it is crucial that you run PCA on your normalized data (which for the sake of this example notebook is median library size normalization + log transformation).\n",
    "\n",
    "Second, we specify n_comps = 100, indicating that we are interested only in the top 100 components.\n",
    "\n",
    "Third, we specify use_highly_variable = True which ensures that PCA is computed only on the highly variable genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41a090d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "612cd48f",
   "metadata": {},
   "source": [
    "To choose the optimal number of principal components, we can visualize the cumulative variance explained:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4a550a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(\n",
    "    range(len(adata.uns[\"pca\"][\"variance_ratio\"])),\n",
    "    np.cumsum(adata.uns[\"pca\"][\"variance_ratio\"]) * 100,\n",
    "    \".-\",\n",
    ")\n",
    "plt.axvline(30, color=\"r\")\n",
    "plt.xlabel(\"Principal Component\", fontsize=14)\n",
    "plt.ylabel(\"% Variance Explained\", fontsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46ebcd8b",
   "metadata": {},
   "source": [
    "As we can see, the increase in variance explained is minimal after about 30 components (as indicated by the red line). So, it is reasonable to perform downstream analysis using 30 principal components.\n",
    "\n",
    "One can argue that it might be sufficient to choose 20 or 25 principal components based on our argument above. But we recommend to be slightly conservative and err on the side of choosing slightly more principal components. However, since there is no absolute way to establish the right number of components, we strongly recommend that you consider evaluating consistency of downstream results (clustering or trajectories) by altering the number of principal components slightly around the chosen value. In the above example, you can see how downstream results change if you alter the number of principal components between 20 to 50."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0577a2c3",
   "metadata": {},
   "source": [
    "For now, we will proceed with 30 components."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eebd5c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm[\"X_pca\"] = adata.obsm[\"X_pca\"][:, 0:30]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b806d85c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm[\"X_pca\"].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dee49a24",
   "metadata": {},
   "source": [
    "Similarly, we want to update the loadings `adata.uns['loadings']` and `adata.varm['PCs']` to reflect only the top 30 principal components."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f32e84c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.varm[\"PCs\"] = adata.varm[\"PCs\"][:, 0:30]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f7839c1",
   "metadata": {},
   "source": [
    "## Note: Genes-Loadings connection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92ae7b99",
   "metadata": {},
   "source": [
    "As we discussed above, we can use the loadings to identify the most influential genes onto each of the PCA.\n",
    "\n",
    "We will sort the gene names by their loadings for the top 30 PCA. First let's create a pandas dataframe of PCA loadings and the corresponding highly variable genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1322ed7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_loadings = pd.DataFrame(\n",
    "    adata.varm[\"PCs\"],\n",
    "    index=adata.var.index,\n",
    "    columns=[\"PC-\" + str(j) for j in range(30)],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f549acc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_loadings.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1ad63c0",
   "metadata": {},
   "source": [
    "Since we ran PCA on HVG, let's make sure that we look at the loadings specific to HVG only. For this we will only consider genes in the HVG."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12827833",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_loadings_hvg = df_loadings.loc[adata.var_names[adata.var[\"highly_variable\"]]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e61ba7c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_loadings_hvg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e947faaa",
   "metadata": {},
   "source": [
    "We can sort the dataframe constructed above for any particular PCA.\n",
    "\n",
    "A note on the sign of the loadings: Since both the positive and negative extreme values of each principal components are treated equally in data interpretion, it stands to reason that we want to look at the absolute value of the loadings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9a825e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.abs(df_loadings_hvg).sort_values(by=\"PC-0\", ascending=False)[\"PC-0\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd1a210e",
   "metadata": {},
   "source": [
    "**Task:** Which genes drive the highest variance in the data?\n",
    "\n",
    "Whether these genes are relevant/important for your study is something you will have to decide. Why might we want to study the loadings? The reason is sometimes we notice that some of the principal components are driven by the expression of mitochondrial genes or ribosomal genes. And depending on the question you are trying to answer it might be of interest to remove the effect of such genes. This way you can identify if the principal components are being influenced by these nuisance genes and if you need to adjust your set of hvg (for example remove mitochondrial genes or ribosomal genes from the list of hvg or add some new set of context specific genes) for downstream computation.\n",
    "\n",
    "**Task:** Plot the genes that drive the highest variance in the data & \"mt_pct_content\", \"log1p_total_counts\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c8604d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.pca(\n",
    "    adata,\n",
    "    color=[   ],\n",
    "    ncols=3,\n",
    "    vmin=\"p1\",\n",
    "    vmax=\"p99\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0757586a",
   "metadata": {},
   "source": [
    "# Nearest neighbor graph"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fa175ac",
   "metadata": {},
   "source": [
    "As we have already mentioned, cellular phenotype space is modeled as a low-dimensional object (called the data manifold) embedded in a high-dimensional space. The expression pattern of genes define the manifold structure. So cells with similar expression profiles are close to each other on the manifold. In other words, the local neighborhood of each cell on the manifold consists of similar cells. This is an important property that must be satisified by any method to accurately capture the structure of the manifold (such as for visualization). In addition, the manifold only depends on the connectivity of the data points - defined via pairwise distances - and not on how the points are arranged in space. Therefore, quantifying cell-cell distance is important to study the manifold structure of the data.\n",
    "\n",
    "In the context of single-cell biology, the data manifold is a characterization of the available space of phenotypes for any cell in that context. Exploring the manifold is therefore equivalent to exploring the possible cellular states, and any defect/perturbation in the molecular processes should reflect as a change in the phenotype landscape. This deviation from normal expression would reflect as a change in characteristics of the data manifold, such as volume or variance or regions of occupancy, analyzing which can help us understand the cause and consequences of the changes.\n",
    "  \n",
    "Understanding the manifold of a high-dimensional data is a problem that pervades several disciplines. As a result, many methods have been proposed to tackle this problem. It is easier to obtain the structure if the underlying manifold is linear - that is, the observed features can be written as a linear combination of the latent variables. Linear methods such as Principal Component Analysis (PCA), discussed above, have been successfully applied in such cases. However, in scRNA-seq data, genes often display non-linear association. The non-linearity is a result of the measured genes themselves being a non-linear combination of the latent variables driving the process, which is in turn a result of the cascades of non-linear interactions, checkpoints and feedback mechanisms a cell has in place. Thus, characterization of the manifold requires us to involve methods that take into account the non-linearity present in single-cell data.\n",
    "\n",
    "To tackle the issue of non-linearity, we aim to characterize the data manifold by first analyzing data locally around every cell, and then combining them. We assume that around each cell, the local neighborhood is roughly linear (or Euclidean), so we will apply linear methods locally and aggregate to get the total picture of the manifold. This notion of analyzing data to characterize the underlying manifold has proven beneficial in single-cell data analysis. From a mathematical point of view, graphs are the perfect tools designed for this. \n",
    "\n",
    "A graph is defined as a set `G = {V,E,W}` consisting of a set of nodes or vertices `V`, a set of edges `E` connecting all or some of the vertices and weights `W` assigned to each of the edges. In the graph representation of single-cell data, each cell is a vertex or node in the graph, two cells are connected if they are similar to each other and the weight between them is the quantification of similarity between them. The similarity between two cells can be obtained by measuring the distance between them. The choice of distance metric, however, is context dependent but due to its ease and useful properties, Euclidean distance seems to be the preferred choice in the single-cell field. The manifold is congruent to a Euclidean space locally around any given point so, it is sensible to use Euclidean distances to compute similarity in the small neighborhood. In particular, for every cell we first find the `k` nearest neighbors (typically `k` is set to < 1% of the total number of cells profiled, a typical choice is `k = 30`) based on Euclidean distance, which will comprise the local neighborhood. Each cell is thus connected to its `k` nearest neighbors by an edge and the constructed graph (called the _k-nearest neighbor graph_) approximates the underlying manifold. The phenotypic distance between any two cells on such graph is then defined by _geodesics_ along the graph, which more accurately captures biological differences between cells.\n",
    "\n",
    "Majority of the visualization, clustering and trajectory inference tools build on top of the above constructed graph. While the thus constructed graph may need to be further adjusted to account for the inherent noise in the data it forms the backbone of most of the tools we will discuss henceforth. We use Scanpy's inbuilt method to construct graph:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cf8135c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(\n",
    "    adata, n_neighbors=30, use_rep=\"X_pca\", metric=\"euclidean\", key_added=\"neighbors_30\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae107eb8",
   "metadata": {},
   "source": [
    "An explanation of the code above:\n",
    "\n",
    "1. We ask for `k = 30` nearest neighbors. This is a typical choice in data analysis but you can compute graphs for multiple values of `k` and study how the downstream result changes (in fact we encourage you to do this to study how your downstream biological interpretations (such as visuals, clustering etc.) alter with slight changes in `k`. For example for this case, you may want to study values of `k` between 20-40.)\n",
    "\n",
    "2. We set `use_rep = 'X_pca'` because we want to compute the k nearest neighbor graph on the above computed PCA. PCA has been observed to be good in capturing global aspects of the data, but to study finer structures of the biological manifold, we recommend building graph on top of the PCA. Furthermore, considering only top principal components has the added advantage of removing noise from the data represented by the lower principal components.\n",
    "\n",
    "3. We set `metric = 'euclidean'` but you are encouraged to explore other metrics such as `'cosine'` or `'correlation'`. The single-cell field still has not agreed upon the best choice of metric or comprehensively identified situations where certain metrics might be better suited than others.\n",
    "\n",
    "4. We set `key_added = 'neighbors_30'` to store our results with the name 'neighbors_30'. This can be a handy tool should you want to compute multiple graphs, as you can simply store multiple results with different names. For more details on this, please see Scanpy API on nearest neighbors. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "455cda3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad4aa170",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsp[\"neighbors_30_distances\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3744b90",
   "metadata": {},
   "source": [
    "The distances (which measure dissimilarity between cells) are represented as a sparse matrix for 11984 cells as shown above. Scanpy also converts the distances into similarities and outputs them as connectivities."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e89a1f05",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsp[\"neighbors_30_connectivities\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8111f249",
   "metadata": {},
   "source": [
    "Note the number of non-zero entries in the matrices are different because the connectivities is a symmetric matrix while distances is not.\n",
    "\n",
    "All major downstream analysis methods utilize thus constructed graph to visualize, cluster or infer trajectories from the data. We begin by exploring the visualization methods and will run UMAP and Force Directed Layout. As such, if you want to change the properties of the neighbor graph (e.g. for a specific visualization or clustering), you would need to re-run the neighbors command."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90df13d2",
   "metadata": {},
   "source": [
    "# Data Visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ad084ad",
   "metadata": {},
   "source": [
    "## UMAP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "317ef41e",
   "metadata": {},
   "source": [
    "Uniform Manifold Approximation and Projection (UMAP) has been applied to visualize and analyze sc-RNAseq data. In summary, UMAP uses the graph we built above and tries to arrange data points on a low-dimensional space (typically in 2D) such that the cross-entropy of the high-dimensional graph weights and low-dimensional graph weights is minimized. The author of UMAP has written a beautiful exposition on how UMAP works, which can be found here: https://bit.ly/2qGhBkk.\n",
    "\n",
    "Note: UMAP preserves high-dimensional distances on the projected space, which suggests that one can do further analysis on UMAP features. However, this should be done with care and we strongly recommend against it. Specifically because we have the tools needed to perform efficient analytic computations (such as clustering, trajectory detection etc.) in the high dimensions itself; thereby precluding the need to do computations on UMAP. For some FAQs on UMAP, please see this page: https://bit.ly/2NfgAxY."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4015ca24",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.umap(adata, neighbors_key=\"neighbors_30\", min_dist=0.1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "57991728",
   "metadata": {},
   "source": [
    "Because we saved our k-nearest neighbor results as `neighbors_30`, we need to specify it as above. We recommend setting min_dist = 0.1 because that was the original default proposed in UMAP algorithm and seems to produce more visually appealing plots. We can do this because the sole purpose of UMAP is to aid in visualization. No quantitative information should be extracted from UMAP (e.g. definite distance/proximity between cells/clusters)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef827b70",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef8a165d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm[\"X_umap\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a13ffc81",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm[\"X_umap\"].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1359c030",
   "metadata": {},
   "source": [
    "Now we can visualize the results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "789b98b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f67495b9",
   "metadata": {},
   "source": [
    "# Visualization examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7404b537",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "132545f5",
   "metadata": {},
   "source": [
    "## Color by metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e452d1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=[\n",
    "        \"log1p_total_counts\",\n",
    "        \"log1p_n_genes_by_counts\",\n",
    "        \"mt_pct_content\",\n",
    "        \"ribo_pct_content\",\n",
    "    ],\n",
    "    ncols=2,\n",
    "    cmap=\"plasma\",\n",
    "    use_raw=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d16f59fd",
   "metadata": {},
   "source": [
    "## Color by gene expression"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b62821f5",
   "metadata": {},
   "source": [
    "We can use the computed UMAP projections to visualize gene expression. This can help us perform some basic sanity checks (if known genes are expressed appropriately), develop hypotheses and aid in cell type annotation (which we will look at next in clustering segment).\n",
    "\n",
    "For now, since we are looking at PBMCs, we will do some sanity checks using established set of genes used to identify various PBMC cell types (list taken from: https://satijalab.org/seurat/archive/v3.0/pbmc3k_tutorial.html)\n",
    "\n",
    "['PTPRC', 'CD3E', 'CD8A', 'CD4', 'IL7R', 'MS4A1', 'CD19', 'LYZ', 'FCGR3A', 'NKG7', 'CST3', 'PPBP']\n",
    "\n",
    "We can make these visuals using Scanpy's inbuilt UMAP plot function. Important note: By default, Scanpy's inbuilt function may rely on `adata.raw` to extract gene expression if you have `adata.raw` saved. So please be careful that you set `use_raw = False` if you have adata.raw set to store the original raw counts as you may rather want to visualize log-normalized counts which is stored in `adata.X`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d67c04b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = [\n",
    "    \"CD3E\",\n",
    "    \"CD8A\",\n",
    "    \"CD4\",\n",
    "    \"IL7R\",\n",
    "    \"MS4A1\",\n",
    "    \"CD19\",\n",
    "    \"LYZ\",\n",
    "    \"FCGR3A\",\n",
    "    \"NKG7\",\n",
    "    \"CST3\",\n",
    "    \"PPBP\",\n",
    "    \"PTPRC\",\n",
    "]\n",
    "sc.pl.umap(adata, color=genes, cmap=\"plasma\", ncols=3, use_raw=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08bdaea0",
   "metadata": {},
   "source": [
    "A note: Sometimes it helps to trim the expression levels of genes to certain percentiles to avoid visualization being biased by a few super-high or super-low values. For this we can utilizie vmin and vmax parameter during plotting as shown below. We will set vmin = 1 percentile of gene expression and vmax = 99 percentile of gene expression, which means min gene expression is going to be 1 percentile of the original expression and analogously for max expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a126a580",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = [\n",
    "    \n",
    "]\n",
    "sc.pl.umap(\n",
    "    adata, color=genes, cmap=\"plasma\", ncols=3, use_raw=False, vmin=\"p1\", vmax=\"p99\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed422174",
   "metadata": {},
   "source": [
    "### Alternate code (more manual)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15cac2c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = [\n",
    "    \n",
    "]\n",
    "\n",
    "fig_nrow = 3\n",
    "fig_ncol = 4\n",
    "\n",
    "fig = plt.figure(figsize=(8 * fig_ncol, 6 * fig_nrow))\n",
    "count = 0\n",
    "for gene_name in genes:\n",
    "    ax = fig.add_subplot(fig_nrow, fig_ncol, count + 1)\n",
    "    col_gene = adata.X.toarray()[:, adata.var_names.get_loc(gene_name)]\n",
    "    im1 = ax.scatter(\n",
    "        adata.obsm[\"X_umap\"][:, 0],\n",
    "        adata.obsm[\"X_umap\"][:, 1],\n",
    "        s=2,\n",
    "        c=col_gene,\n",
    "        cmap=\"plasma\",\n",
    "    )\n",
    "    ax.set_xlabel(\"UMAP-1\", fontsize=14)\n",
    "    ax.set_ylabel(\"UMAP-2\", fontsize=14)\n",
    "    ax.set_title(\"UMAP colored by \" + gene_name, fontsize=14)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    fig.colorbar(im1)\n",
    "    count = count + 1\n",
    "\n",
    "# To save, please uncomment the following line of code:\n",
    "# fig.savefig(output_directory, 'UMAP_colored_by_genes.png', dpi = 150, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d74cd473",
   "metadata": {},
   "source": [
    "Such visualizations can help perform data analysis better. You may already be able to draw many hypotheses and make assessment of the data using such visualizations.\n",
    "\n",
    "Similarly, alternate code to trim the gene expression at 1 percentile and 99 percentile."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29a53594",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = [\n",
    "   \n",
    "]\n",
    "\n",
    "fig_nrow = 3\n",
    "fig_ncol = 4\n",
    "\n",
    "fig = plt.figure(figsize=(8 * fig_ncol, 6 * fig_nrow))\n",
    "count = 0\n",
    "for gene_name in genes:\n",
    "    ax = fig.add_subplot(fig_nrow, fig_ncol, count + 1)\n",
    "    col_gene = adata.X.toarray()[:, adata.var_names.get_loc(gene_name)]\n",
    "    im1 = ax.scatter(\n",
    "        adata.obsm[\"X_umap\"][:, 0],\n",
    "        adata.obsm[\"X_umap\"][:, 1],\n",
    "        s=2,\n",
    "        c=col_gene,\n",
    "        cmap=\"plasma\",\n",
    "        vmin=np.percentile(col_gene, 1),\n",
    "        vmax=np.percentile(col_gene, 99),\n",
    "    )\n",
    "    ax.set_xlabel(\"UMAP-1\", fontsize=14)\n",
    "    ax.set_ylabel(\"UMAP-2\", fontsize=14)\n",
    "    ax.set_title(\"UMAP colored by \" + gene_name, fontsize=14)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    fig.colorbar(im1)\n",
    "    count = count + 1\n",
    "\n",
    "# To save, please uncomment the following line of code:\n",
    "# fig.savefig(output_directory, 'UMAP_colored_by_genes.png', dpi = 150, bbox_inches = 'tight')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3d133689",
   "metadata": {},
   "source": [
    "You can also visualize established gene signatures (from published papers or bulk signatures) onto the computed UMAP. For example, you may be familiar with Hallmark or Reactome or GO genesets. You can load any of your favorite pathway and visualize the expression using the following code as a guideline. For illustration, we have provided you with an example geneset called geneset.txt and we will use that to visualize the expression (taken from: https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/HALLMARK_G2M_CHECKPOINT.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38f8302e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the file\n",
    "geneset = pd.read_csv(data_path + \"HALLMARK_G2M_CHECKPOINT.v2023.2.Hs.grp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3bd1b59",
   "metadata": {},
   "outputs": [],
   "source": [
    "geneset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "376e410f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = geneset[\"HALLMARK_G2M_CHECKPOINT\"]\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46ca0ed0",
   "metadata": {},
   "outputs": [],
   "source": [
    "type(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1050654",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute a gene signature score using Scanpy's inbuilt function\n",
    "sc.tl.score_genes(\n",
    "    adata, genes, score_name=\"G2_M_checkpoint\", random_state=0, use_raw=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b14de03",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7816c838",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=\"G2_M_checkpoint\", cmap=\"plasma\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19c5cefc",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=\"G2_M_checkpoint\", cmap=\"plasma\", vmin=\"p1\", vmax=\"p99\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e67f88bd",
   "metadata": {},
   "source": [
    "# Clustering"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "5e84a1cf",
   "metadata": {},
   "source": [
    "One of the primary applications of scRNA-seq has been in the identification of novel and rare cell types. In contrast to bulk measurements, which obscure the cell-cell heterogeneity, single-cell level analysis offers an unprecedented precision to classify cells into different types.\n",
    "\n",
    "To think about it computationally: The data presents us with a bunch of objects (individual cells) that have a certain set of features (genes), and we would like to classify the objects into distinct groups or categories each consisting of similar cells. In our case, these distinct groups or categories can be biologically interpreted as cell types. Therefore, in essence the problem of identifying cell types (novel, rare or known) boils down to a problem of clustering.\n",
    "\n",
    "Performing Modularity optimization by Leiden community detection on the k-nearest-neighbour graph of cells has become an established practice in scRNA-seq analysis. Thus, this is the method of choice in this tutorial as well.\n",
    "\n",
    "Here, we perform clustering at two resolutions. Investigating several resolutions allows us to select a clustering that appears to capture the main clusters in the visualization and can provide a good baseline for further subclustering of the data to identify more specific substructure.\n",
    "\n",
    "Clustering is performed on the highly variable gene data, dimensionality reduced by PCA, and embedded into a KNN graph."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3ccb4caa",
   "metadata": {},
   "source": [
    "Compute a `leiden` clustering with two different resolutions (`0.5` and `1.5`). Compare the clusterings in a table and visualize the clustering in an embedding."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1bd40cc2",
   "metadata": {},
   "source": [
    "## Leiden clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e423068a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.leiden(\n",
    "    adata,\n",
    "    resolution=1.5,\n",
    "    random_state=0,\n",
    "    neighbors_key=\"neighbors_30\",\n",
    "    key_added=\"leiden_r1.5\",\n",
    ")\n",
    "sc.tl.leiden(\n",
    "    adata,\n",
    "    resolution=0.5,\n",
    "    random_state=0,\n",
    "    neighbors_key=\"neighbors_30\",\n",
    "    key_added=\"leiden_r0.5\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d674a5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06c7084c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.crosstab(adata.obs[\"leiden_r0.5\"], adata.obs[\"leiden_r1.5\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b862109",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize the clustering\n",
    "sc.pl.umap(adata, color=[\"leiden_r1.5\", \"leiden_r0.5\"], wspace=0.6)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b59fd078",
   "metadata": {},
   "source": [
    "## Investigate % mito content, library size, number of genes expressed in each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c218e682",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3f472e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_temp = pd.DataFrame(\n",
    "    {\n",
    "        \"leiden\": adata.obs[\"leiden_r1.5\"],\n",
    "        \"%-Mito\": adata.obs[\"mt_pct_content\"],\n",
    "        \"library_size\": adata.obs[\"log1p_total_counts\"],\n",
    "        \"n_genes_per_cell\": adata.obs[\"log1p_n_genes_by_counts\"],\n",
    "    },\n",
    "    index=adata.obs.index,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c14f353",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5444359",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "\n",
    "fig = plt.figure(figsize=(8 * 2, 6 * 3))\n",
    "ax = fig.add_subplot(3, 1, 1)\n",
    "sns.boxplot(\n",
    "    x=\"leiden\",\n",
    "    y=\"%-Mito\",\n",
    "    hue=\"leiden\",\n",
    "    data=df_temp,\n",
    "    ax=ax,\n",
    "    palette=sc.pl.palettes.zeileis_28,\n",
    ")\n",
    "ax.grid(\"on\")\n",
    "ax.legend().set_visible(False)\n",
    "\n",
    "ax = fig.add_subplot(3, 1, 2)\n",
    "sns.boxplot(\n",
    "    x=\"leiden\",\n",
    "    y=\"library_size\",\n",
    "    hue=\"leiden\",\n",
    "    data=df_temp,\n",
    "    ax=ax,\n",
    "    palette=sc.pl.palettes.zeileis_28,\n",
    ")\n",
    "ax.grid(\"on\")\n",
    "ax.legend().set_visible(False)\n",
    "\n",
    "ax = fig.add_subplot(3, 1, 3)\n",
    "sns.boxplot(\n",
    "    x=\"leiden\",\n",
    "    y=\"n_genes_per_cell\",\n",
    "    hue=\"leiden\",\n",
    "    data=df_temp,\n",
    "    ax=ax,\n",
    "    palette=sc.pl.palettes.zeileis_28,\n",
    ")\n",
    "ax.grid(\"on\")\n",
    "ax.legend(loc=\"lower left\", ncol=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "343adf3b",
   "metadata": {},
   "source": [
    "### Highlight clusters of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a10f27e",
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters_interest = [\"\"]  # explore\n",
    "colors = [\"red\"]\n",
    "plt.scatter(adata.obsm[\"X_umap\"][:, 0], adata.obsm[\"X_umap\"][:, 1], s=1, color=\"gray\")\n",
    "for counter, item in enumerate(clusters_interest):\n",
    "    cells_in_cluster = adata.obs[\"leiden_r1.5\"] == item\n",
    "    plt.scatter(\n",
    "        adata.obsm[\"X_umap\"][cells_in_cluster, 0],\n",
    "        adata.obsm[\"X_umap\"][cells_in_cluster, 1],\n",
    "        s=1,\n",
    "        color=colors[counter],\n",
    "        label=\"Cluster-\" + str(item),\n",
    "    )\n",
    "\n",
    "plt.legend(markerscale=5)\n",
    "plt.axis(\"off\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "6193403e",
   "metadata": {},
   "source": [
    "### Remove the clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67942857",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_clean = adata[~np.isin(adata.obs[\"leiden_r1.5\"], clusters_interest), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f19977ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_clean"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c3936a4",
   "metadata": {},
   "source": [
    "# Re-do pre-processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d5c1088",
   "metadata": {},
   "outputs": [],
   "source": [
    "# HVG\n",
    "sc.pp.highly_variable_genes(\n",
    "    adata_clean, layer=\"raw_data\", n_top_genes=4000, flavor=\"seurat_v3\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79e942d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PCA\n",
    "sc.tl.pca(adata_clean, n_comps=30, use_highly_variable=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16c480b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nearest Neighbors\n",
    "sc.pp.neighbors(adata_clean, n_neighbors=30, use_rep=\"X_pca\", key_added=\"neighbors_30\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c3aa095",
   "metadata": {},
   "outputs": [],
   "source": [
    "# UMAP\n",
    "sc.tl.umap(adata_clean, neighbors_key=\"neighbors_30\", min_dist=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b786c70",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clustering\n",
    "sc.tl.leiden(adata_clean, resolution=1, random_state=0, neighbors_key=\"neighbors_30\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd07c7bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_clean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12e3f2b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_clean, color=\"leiden\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "01f01564",
   "metadata": {},
   "source": [
    "# Marker genes and cluster annotation "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "581e8bd8",
   "metadata": {},
   "source": [
    "To annotate the clusters we obtained, we find genes that are up-regulated in the cluster compared to all other clusters (marker genes). This differential expression test is performed by a *Welch t-test with overestimated variance* to be conservative. This is the default in `scanpy`. The test is automatically performed on the `.raw` data set, which is uncorrected and contains all genes. All genes are taken into account, as any gene may be an informative marker.\n",
    "\n",
    "Annotation can be performed in an manual fashion or automated. \n",
    "* Manual annotation: based on the expression of marker genes, consultation with experts \n",
    "* Automated annotation: the development of large-scale atlases of different tissues allows to map the datasets to references. They are based on different principles, sometimes requiring pre-defined sets of markers, other times trained on pre-existing full scRNA-seq datasets.See also several reviews [Pasquini et al., 2021](10.1016/j.csbj.2021.01.015), [Abdelaal et al., 2019](https://doi.org/10.1186/s13059-019-1795-z) for a more elaborate discussion of automated annotation methods.\n",
    "\n",
    "The resulting annotations can be of varying quality. It is therefore important to regard these methods as a starting point rather than an end-point of the annotation process."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4db50aa2",
   "metadata": {},
   "source": [
    "Compute the differential expression profile for each cluster with `rank_genes_groups` and visualize the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89df79a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_clean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "046afa6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.log1p(adata_clean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa7f3fcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate marker genes\n",
    "sc.tl.rank_genes_groups(adata_clean, groupby=\"leiden\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8196e894",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot marker genes\n",
    "sc.pl.rank_genes_groups(adata_clean, fontsize=12)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f8a338e3",
   "metadata": {},
   "source": [
    "Here, we observe potentially characteristic gene expression patterns.\n",
    "\n",
    "Furthermore, the score itself is not interpretable in terms of specificity and significance in the case of clustering, because the clusters were previously defined as a group of cells being different from the rest. Therefore, we compare a group that is a priori different to the rest and the resulting scores (or p-values) are inflated. Furthermore, the smaller a cluster is, the smaller is the observed score, unless a gene is very specific to the cluster. Typically, we may find marker genes in the gene lists of the `rank_genes_groups` test, but not all marker genes have a high expression level."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fe465efe",
   "metadata": {},
   "source": [
    "## Wilcoxon test"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f727a965",
   "metadata": {},
   "source": [
    "Seurat has a FindMarkers function (https://satijalab.org/seurat/reference/findmarkers) to compute differentially expressed genes, and by default it performs wilcoxon test for a gene in two different groups (typically on z-scored/scaled data). We will use the rank_genes_groups function in Scanpy to do the same."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "748dddcb",
   "metadata": {},
   "source": [
    "On log normalized data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3938e3d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(\n",
    "    adata_clean,\n",
    "    groupby=\"leiden\",\n",
    "    groups=\"all\",\n",
    "    reference=\"rest\",\n",
    "    key_added=\"on_norm_log\",\n",
    "    method=\"wilcoxon\",\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b0b1f2bf",
   "metadata": {},
   "source": [
    "Example output for Cluster 0:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9a8d52b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_wt = sc.get.rank_genes_groups_df(adata_clean, group=str(0), key=\"on_norm_log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f1a7a93",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_wt"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fa6d18c8",
   "metadata": {},
   "source": [
    "Get the results for each cluster and save:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57530408",
   "metadata": {},
   "outputs": [],
   "source": [
    "for cluster in np.unique(adata_clean.obs[\"leiden\"]):\n",
    "    df = sc.get.rank_genes_groups_df(adata_clean, group=str(cluster), key=\"on_norm_log\")\n",
    "    df.to_csv(\"path/to/save/wilcoxon_cluster_\" + str(cluster) + \"_vs_rest.csv\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e7a74c68",
   "metadata": {},
   "source": [
    "## Other approaches:"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e5807d1e",
   "metadata": {},
   "source": [
    "If you wish to test other DEA approaches, please consider the following suggestings:\n",
    "\n",
    "1) pseudobulk based pyDESeq2 analysis  \n",
    "\n",
    "2) Memento: https://www.biorxiv.org/content/10.1101/2022.11.09.515836v1, https://github.com/yelabucsf/scrna-parameter-estimation"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "85589333",
   "metadata": {},
   "source": [
    "## Explore data using prior knowledge"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4ed11a8e",
   "metadata": {},
   "source": [
    "When it comes to cluster annotation, we usually have to tap into prior knowledge of the cell type. Depending on the data set, this may involve extensive literature search.\n",
    "\n",
    "In the case of PBMCs, we may refer to several studies and single-cell RNA-sequencing data analysis tutorials to extract marker gene sets. \n",
    "\n",
    "The following list is extracted from the Seurat tutorial on PBMCs.\n",
    "\n",
    "|Marker Gene|Cell Type|\n",
    "|---------|-------|\n",
    "|IL7R|CD4 T cells|\n",
    "|CD14, LYZ|CD14+ Monocytes|\n",
    "|MS4A1|B cells|\n",
    "|CD8A|CD8 T cells|\n",
    "|FCGR3A, MS4A7|FCGR3A+ Monocytes|\n",
    "|GNLY, NKG7|NK cells|\n",
    "|FCER1A, CST3|Dendritic Cells|\n",
    "|PPBP|Megakaryocytes|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4156ccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_genes = [\n",
    "\n",
    "]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "582afb12",
   "metadata": {},
   "source": [
    "**Tasks:** Annotate the clusters. \n",
    "Check briefly, if all marker genes are present in the dataset and visualise the marker genes in a UMAP (or another visualisation of your choice).\n",
    "You can use auxiliary plots like `matrixplot`, `dotplot`, `heatmap` or `violin` plots or coloring an embedding (e.g. UMAP) by the marker genes."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "7be487ba",
   "metadata": {},
   "source": [
    "Let us check if the marker genes are expressed in our dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7621855d",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.in1d(marker_genes, adata_clean.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40cdae74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5970832",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata=, \n",
    "              var_names =,\n",
    "              groupby=, \n",
    "              use_raw=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5c8e7f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.heatmap(adata=, \n",
    "              var_names=, \n",
    "              figsize=(5,10),\n",
    "              groupby=, \n",
    "              use_raw=False, vmin=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf38f0f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.matrixplot(adata=, \n",
    "                 var_names=,\n",
    "                 groupby=, \n",
    "                 use_raw=False, vmin=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b7fb300",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.stacked_violin(adata = ,\n",
    "                     var_names = , \n",
    "                     groupby=, \n",
    "                     use_raw=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "7f226754",
   "metadata": {},
   "source": [
    "Annotate clusters and create a new covariate.\n",
    "\n",
    "\n",
    "|Cluster ID|Marker Gene|Cell Type|\n",
    "|---------|-------|-------|\n",
    "||IL7R|CD4 T cells|\n",
    "||CD14, LYZ|CD14+ Monocytes|\n",
    "||MS4A1|B cells|\n",
    "||CD8A|CD8 T cells|\n",
    "||FCGR3A, MS4A7|FCGR3A+ Monocytes|\n",
    "||GNLY, NKG7|NK cells|\n",
    "||FCER1A, CST3|Dendritic Cells|\n",
    "||PPBP|Megakaryocytes|"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "d2ac1d46",
   "metadata": {},
   "source": [
    "Use the `pandas` data frame functionality to rename your clusters and visualize your annotation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "568ff962",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_clean.obs[\"annotated\"] = adata_clean.obs[\"louvain_r1.5\"].cat.add_categories(\n",
    "    [\n",
    "\n",
    "    ]\n",
    ")\n",
    "\n",
    "adata_clean.obs[\"annotated\"][\n",
    "    np.in1d(adata_clean.obs[\"annotated\"], [])  # add cluster name here (as string)\n",
    "] = \"CD4 T cells\"\n",
    "adata_clean.obs[\"annotated\"][\n",
    "    np.in1d(adata_clean.obs[\"annotated\"], [])\n",
    "] = \"CD14+ Monocytes\"\n",
    "adata_clean.obs[\"annotated\"][np.in1d(adata_clean.obs[\"annotated\"], [])] = \"B cells\"\n",
    "adata_clean.obs[\"annotated\"][np.in1d(adata_clean.obs[\"annotated\"], [])] = \"CD8 T cells\"\n",
    "adata_clean.obs[\"annotated\"][\n",
    "    np.in1d(adata_clean.obs[\"annotated\"], [])\n",
    "] = \"FCGR3A+ Monocytes\"\n",
    "adata_clean.obs[\"annotated\"][np.in1d(adata_clean.obs[\"annotated\"], [])] = \"NK cells\"\n",
    "adata_clean.obs[\"annotated\"][\n",
    "    np.in1d(adata_clean.obs[\"annotated\"], [])\n",
    "] = \"Dendritic cells\"\n",
    "adata_clean.obs[\"annotated\"][\n",
    "    np.in1d(adata_clean.obs[\"annotated\"], [])\n",
    "] = \"Megakaryocytes\"\n",
    "\n",
    "# remove unused categories from annotation\n",
    "adata_clean.obs[\"annotated\"] = adata.obs[\"annotated\"].cat.remove_unused_categories()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f81f7647",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check how many cells per cell type annotated\n",
    "adata_clean.obs[\"annotated\"].value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "020c79d3",
   "metadata": {},
   "source": [
    "**Task:** Visualise your annotation on a UMAP as well as in a `matrixplot`, `dotplot`, `heatmap` or `violin` plots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "171ec3a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata = , color='annotated', legend_loc='on data', title='', frameon=False)\n",
    "sc.pl.umap(adata = , color='annotated',  title='', frameon=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "874b4a91",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata=, \n",
    "              var_names =,\n",
    "              groupby=, \n",
    "              use_raw=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8367d93",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.heatmap(adata=, \n",
    "              var_names=, \n",
    "              figsize=(5,10),\n",
    "              groupby=, \n",
    "              use_raw=False, vmin=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa59232d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.matrixplot(adata = , \n",
    "                 var_names = , \n",
    "                 groupby= , \n",
    "                 use_raw=False, vmin=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee3f68bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.stacked_violin(adata=   , \n",
    "                     var_names=   , \n",
    "                     groupby=    , \n",
    "                     use_raw=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f1d2f3a",
   "metadata": {},
   "source": [
    "# Save results for future"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aacb5713",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_clean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc957bdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_clean = adata_clean.write_h5ad(data_path + \"processed/annotated_data.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cab57c91",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "307.188px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
