{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lecture 9: Spatial Analysis and Spatial Structure\n",
    "\n",
    "**Date:** January 2, 2026 | **Time:** 60 minutes\n",
    "\n",
    "## Learning Objectives\n",
    "- Process and analyze spatial transcriptomics data using squidpy\n",
    "- Compute spatial statistics and neighborhood enrichment\n",
    "- Identify spatially variable genes\n",
    "- Characterize spatial niches\n",
    "\n",
    "---\n",
    "\n",
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import squidpy as sq\n",
    "import numpy as np\n",
    "\n",
    "sc.settings.verbosity = 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1: Loading Spatial Data (20 points)\n",
    "\n",
    "### Instructions\n",
    "1. Load Visium mouse brain dataset: `sq.datasets.visium_hne_adata()`\n",
    "2. Visualize spatial coordinates\n",
    "3. Show gene expression in spatial context\n",
    "4. Plot total counts in tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Load and visualize spatial data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2: Spatial Neighborhood Analysis (30 points)\n",
    "\n",
    "### Instructions\n",
    "1. Build spatial graph using `sq.gr.spatial_neighbors`\n",
    "2. Perform clustering on spatial data\n",
    "3. Compute neighborhood enrichment\n",
    "4. Visualize which cell types are spatially co-localized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Analyze spatial neighborhoods\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3: Spatially Variable Genes (30 points)\n",
    "\n",
    "### Instructions\n",
    "1. Identify spatially variable genes using Moran's I\n",
    "2. Visualize top SVGs in tissue\n",
    "3. Compare with highly variable genes\n",
    "4. Interpret biological meaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Find spatially variable genes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 4: Spatial Niches (20 points)\n",
    "\n",
    "### Instructions\n",
    "1. Define spatial niches based on expression\n",
    "2. Identify niche-specific markers\n",
    "3. Visualize niches in tissue\n",
    "4. Compare with anatomical structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Characterize spatial niches\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
